### DISCUSSION TITLE
HW4: LIA Synthesizer

### DISCUSSION THREAD

**Discussion by spearo2** [prosyslab-classroom/cs424-program-reasoning#434] [Announcements]
안녕하세요! HW4: LIA Synthesizer가 공개되었습니다.
본 과제의 목표는 선형 정수 산술(Linear Integer Arithmetic)로 이루어진 프로그램을 수업에서 배운 top-down/bottom-up 탐색 방식을 사용하여 합성하는것 입니다.
아래 링크를 통해 과제를 시작해 주세요. [https://classroom.github.com/a/iYOCW21q](https://classroom.github.com/a/iYOCW21q)

** 과제를 완료하시면 꼭! 마지막 커밋을 리포지토리에 푸시해 주시고, [Gradescope](https://www.gradescope.com/courses/840493) 에 깃허브 리포지토리를 통해 제출해 주세요.
과제 마감은 11.21(목) 23:59 입니다. 11.23(토) 23:59 까지 제출은 가능하지만 마감 기한 후에는 수업의 [채점 규정](https://github.com/prosyslab-classroom/cs424-program-reasoning?tab=readme-ov-file#%EC%88%99%EC%A0%9C-homework)에 따른 감점이 있습니다.

본 과제에서는 이전과 다른 채점 방식이 적용됩니다.
본 과제는 제출물의 정확도뿐만 아니라 효율성 또한 채점합니다.
자세한 채점 방식은 다음과 같습니다.

* 학생분들은 Gradescope에 계속해서 구현하신 합성기를 제출하실 수 있습니다. Gradescope는 제출물의 top-down과 bottom-up 방식의 합성을 20개의 테스트케이스로 채점합니다. (이중, 5개는 템플릿 코드에 주어진 예시입니다.)
* 본과제는 이전 과제들과 같이 커버리지를 채점 기준으로 사용하지 않습니다. 대신에, 최소 5개의 개인 테스트케이스를 직접 작성하여 템플릿 코드의 `mytest` 디렉토리에 추가해 주셔야 합니다. 
* `mytest`디렉토리에 추가된 테스트 케이스가 5개보다 적거나 테스트케이스의 문법에 문제가 있다면 기존 커버리지의 방식과 같이 0점으로 처리됩니다.
* 개인 테스트케이스 작성은 `test`디렉토리에 있는 `*.sl`파일을 참고하여 주세요. (`mytest` 디렉토리에는 `dune`을 추가하지 않으셔도 됩니다.)
* 과제가 마감되면 학생분들이 제출하신 테스트케이스 중에 임의로 5개를 선정하여 제출된 과제를 추가로 채점하게 됩니다. 추가로 채점된 부분은 기존 순위표의 성적과 합산되어 게시됩니다.
* 각 테스트케이스의 점수는 다음과 같이 계산됩니다. 
if Timeout then -10 else (60 - time) + (10 - log₂(size))
    * time 은 각 테스트케이스를 합성할때 소요되는 시간입니다(초).
    * Timeout은 문제당 1분으로 설정되어 있습니다.
    * size 는 합성된 프로그램이 가진 터미널(terminal)의 갯수 입니다.
    * 합성된 프로그램이 작을수록 높은 점수를 받게 됩니다.
    * 총점은 각 테스트케이스에서 계산된 점수의 합산입니다.
* 이렇게 산출된 점수를 0~100의 범위로 환산하여 점수가 게시됩니다.

학생분들의 동기부여를 위해 순위표에 작년과 제작년에 동일과제에서 우수한 성적을 거둔 학생의 점수를 게시하였습니다.

감사합니다.

=====================================================================================

Hi all, this is an announcement about Homework 4: LIA Synthesizer.

The goal of this assignment is to synthesize programs consisting of Linear Integer Arithmetic (LIA) using the top-down/bottom-up enumerative search methods learned in class.

Here is the GitHub classroom link for this homework:
[https://classroom.github.com/a/iYOCW21q](https://classroom.github.com/a/iYOCW21q)

After completing the assignment, please make sure to PUSH your final commit to the repository and submit it via GitHub repository on [Gradescope](https://www.gradescope.com/courses/840493) in order to get graded.
* Due date: 11.21 (Thursday) 23:59
* Late submission: Accepted until 11.23 (Saturday) 23:59 (with point deductions according to the course's grading policy)

Important Notes:
In this assignment, a different grading method from previous assignments will be applied. This assignment will grade not only the correctness of your submission but also its efficiency. The detailed grading method is as follows:
* Gradescope Submissions: You can continue to submit your implemented synthesizers to Gradescope. Gradescope will grade the correctness of your submission for top-down and bottom-up methods with 20 test cases. (5 of them are given as examples in the template code)
* Custom Test Cases: This assignment does not use coverage as a grading criterion like previous assignments. Instead, you must write at least 5 personal custom test cases and add them to the `mytest` directory in the template code.
    * If there are fewer than 5 test cases added to the mytest directory or if there is a problem with the syntax of the test cases, your submission will be graded as zero points as in the previous coverage method.
    * For writing personal test cases, please refer to the *.sl files in the `test` directory. (You do not need to add dune to the `mytest` directory.)
* Additional Grading After Deadline: After the assignment deadline, we will randomly select 5 test cases from the test cases you submitted and additionally grade the submitted assignments. The additionally graded parts will be added to the existing ranking table scores.
* Scoring for Each Test Case: The score for each test case is calculated as follows: 
if Timeout then -10 else (60 - time) + (10 - log₂(size))
     * time denotes the elapsed time (sec) for synthesis for each problem. 
     * We have set the timeout to one minute per problem. 
     * size indicates the number of terminals in your synthesized program. 
     * We will give a higher score to the smaller program. 
     * The total score will be the sum of the scores of each test case.
* The calculated scores will be scaled to a range of 0~100 and posted to the leaderboard.

* Motivation: For your motivation, we have posted the scores of students who achieved excellent grades on the same assignment in last year and the year before.

Hope you have fun!

Thanks.

**Comment by Akais24** [comment]
제공된 링크의 과제가 HW4가 아니라 HW2인 것 같습니다
텍스트는 정상인데 하이퍼 링크가 잘못 걸려 있는 것 같습니다

---

It appears that the assignment provided in the link is HW2, not HW4
The text seems to be correct, but it appears that the hyperlink is incorrectly attached.

**Reply by sujin0529** [reply]
하이퍼 링크 수정했습니다.
감사합니다.

---
I fixed hyperlink.
Thanks.
